astropy/units/format/base.py

```diff
@@ -270,24 +296,35 @@ def _invalid_unit_error_message(cls, unit: str) -> str:
         )
 
     @classmethod
-    def _decompose_to_known_units(cls, unit: CompositeUnit | NamedUnit) -> UnitBase:
+    def _decompose_to_known_units(
+        cls,
+        unit: CompositeUnit | NamedUnit,
+        deprecations: Literal["silent", "warn", "raise"] = "warn",
+    ) -> UnitBase:
         """
         Partially decomposes a unit so it is only composed of units that
         are "known" to a given format.
         """
         if isinstance(unit, CompositeUnit):
             return CompositeUnit(
                 unit.scale,
-                [cls._decompose_to_known_units(base) for base in unit.bases],
+                [
+                    cls._decompose_to_known_units(base, deprecations)
+                    for base in unit.bases
+                ],
                 unit.powers,
                 _error_check=False,
             )
         if isinstance(unit, NamedUnit):
             try:
-                return cls._validate_unit(unit._get_format_name(cls.name))
-            except ValueError:
-                if isinstance(unit, Unit):
-                    return cls._decompose_to_known_units(unit._represents)
+                return cls._validate_unit(
+                    unit._get_format_name(cls.name), deprecations=deprecations
+                )
+            except ValueError as err:
+                if isinstance(unit, Unit) and not str(err).startswith(
+                    "invalid deprecation handling option"
+                ):
```

Comment:
If the user has provided an invalid `deprecations` value then `_validate_unit()` should raise a `ValueError` about that. But currently `_validate_unit()` also raises a `ValueError` if the unit is completely invalid, so the code here has to inspect the error message to find out what the cause of the `ValueError` was. With #18593 an invalid unit will cause a `KeyError` instead of a `ValueError`, so we will not have to inspect the error message here.